### DISCUSSION TITLE
8.13 build fails with undefined reference to `dlsym'

### DISCUSSION THREAD

**Discussion by adityapatadia** [libvips/libvips#3000] [Q&A]
Here is full error that I get when I try to build libvips.

```
#18 56.79 [362/364] Linking target libvips/libvips.so.42.15.0
#18 56.79 FAILED: libvips/libvips.so.42.15.0 
#18 56.79 c++  -o libvips/libvips.so.42.15.0 libvips/libvips.so.42.15.0.p/meson-generated_.._include_vips_enumtypes.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 -shared -fPIC -Wl,--start-group -Wl,-soname,libvips.so.42 -Wl,--whole-archive libvips/foreign/libnsgif/libnsgif.a libvips/foreign/libforeign.a libvips/arithmetic/libarithmetic.a libvips/resample/libresample.a libvips/colour/libcolour.a libvips/conversion/libconversion.a libvips/convolution/libconvolution.a libvips/freqfilt/libfreqfilt.a libvips/histogram/libhistogram.a libvips/draw/libdraw.a libvips/iofuncs/libiofuncs.a libvips/morphology/libmorphology.a libvips/mosaicing/libmosaicing.a libvips/create/libcreate.a -Wl,--no-whole-archive -O3 -march=cascadelake -mno-avx512f -fPIC -Wl,-z,nodelete /usr/lib/x86_64-linux-gnu/libglib-2.0.so /usr/lib/x86_64-linux-gnu/libgio-2.0.so /usr/lib/x86_64-linux-gnu/libgobject-2.0.so /usr/lib/x86_64-linux-gnu/libexpat.so -pthread -lm /usr/lib/x86_64-linux-gnu/libz.so /usr/lib/x86_64-linux-gnu/libgsf-1.so /usr/lib/x86_64-linux-gnu/libxml2.so /usr/lib/x86_64-linux-gnu/libfftw3.so /usr/lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/libimagequant.so /usr/lib/x86_64-linux-gnu/libexif.so /usr/lib/x86_64-linux-gnu/libjpeg.so /usr/local/lib/libspng.so /usr/lib/libwebp.so /usr/lib/libwebpmux.so /usr/lib/libwebpdemux.so /usr/lib/x86_64-linux-gnu/libpangocairo-1.0.so /usr/lib/x86_64-linux-gnu/libpango-1.0.so /usr/lib/x86_64-linux-gnu/libharfbuzz.so /usr/lib/x86_64-linux-gnu/libcairo.so /usr/lib/x86_64-linux-gnu/libpangoft2-1.0.so /usr/lib/x86_64-linux-gnu/libfontconfig.so /usr/lib/x86_64-linux-gnu/libfreetype.so /usr/lib/x86_64-linux-gnu/libtiff.so /usr/local/lib/librsvg-2.a /usr/lib/x86_64-linux-gnu/libgdk_pixbuf-2.0.so /usr/lib/x86_64-linux-gnu/libcairo-gobject.so /usr/lib/x86_64-linux-gnu/liblcms2.so /usr/lib/x86_64-linux-gnu/liborc-0.4.so /usr/local/lib/libpdfium.so /usr/local/lib/libheif.so -Wl,--end-group -Wl,-rpath,/usr/local/lib -Wl,-rpath-link,/usr/local/lib
#18 56.79 /usr/bin/ld: /usr/local/lib/librsvg-2.a(std-8f1929c73c3f8167.std.4fd68eda-cgu.0.rcgu.o): in function `std::sys::unix::weak::DlsymWeak<F>::initialize':
#18 56.79 (.text.unlikely._ZN3std3sys4unix4weak18DlsymWeak$LT$F$GT$10initialize17h9931962ec31da8f6E+0x3b): undefined reference to `dlsym'
#18 56.79 collect2: error: ld returned 1 exit status
#18 56.79 ninja: build stopped: subcommand failed.
```

I searched online and someone suggested to add some flags so I added
```
${LINUX:+-Dcpp_link_args="$LDFLAGS -Wl,-Bsymbolic-functions -Wl,--no-as-needed -ldl"} 
```

But it's still not building and failing with same error.

**Comment by jcupitt** [comment]
Hi @adityapatadia,

libvips doesn't use `dlsym` directly, it uses the gmodule dynamic library wrapper. If linking is failing I guess something is messed up in your glib install. 

What platform are you using, which glib version are you using, how did you set up libvips, etc.?

**Comment by adityapatadia** [comment]
We are using `node:16.16.0-bullseye-slim` docker image which gets its glib from bullseye repo I guess. We are self-compiling libvips with some required libraries fetched from debian repos and libraries like librsvg, libwebp, mozjpeg, aom, libspng are compiled. We are also adding PDFium as per instructions.

Everything else is same and worked fine till 8.12.2 but upgrading to 8.13 and using meson fails now.

**Comment by kleisauke** [comment]
IIRC, you were using a custom build script based on https://github.com/lovell/sharp-libvips. Are you building librsvg with `CARGO_PROFILE_RELEASE_LTO=true`?
https://github.com/lovell/sharp-libvips/blob/e091d07a6e15d961580d012ea9ff376e8169870b/build/lin.sh#L78

Rust's LTO optimizations probably doesn't make librsvg dependent on `dlopen` and `dlsym`. But if LTO cannot be used (perhaps due to https://github.com/rust-lang/rust/issues/44322), then you could try to make librsvg pkg-config file aware of the `libdl` dependency by doing:
```bash
sed -i '/^Libs:/ s/$/ -ldl/' librsvg.pc.in
```

Just before configuring librsvg. Another option is to just build librsvg as a shared library, just like the other dependencies.

**Reply by adityapatadia** [reply]
This one worked like a charm :)

**Comment by adityapatadia** [comment]
I am fine to build it as shared library. I would appreciate if you can share steps to do it. Currently this is how we build librsvg.

```bash
RUN without_patch() { echo "${1%.[[:digit:]]*}";}; wget -q -O librsvg-${LIBRSVG_VERSION}.tar.gz https://download.gnome.org/sources/librsvg/$(without_patch $LIBRSVG_VERSION)/librsvg-${LIBRSVG_VERSION}.tar.xz && \
  ## first we are installing rust and cargo
  curl https://sh.rustup.rs -sSf | sh -s -- -y && . $HOME/.cargo/env && \
  \
  tar -xf librsvg-${LIBRSVG_VERSION}.tar.gz && \
  cd librsvg-${LIBRSVG_VERSION} && \
  ## Compiling librsvg itself
  sed -i'.bak' "s/^\(Requires:.*\)/\1 cairo-gobject pangocairo/" librsvg.pc.in && \
  # LTO optimization does not work for staticlib+rlib compilation
  sed -i'.bak' "s/, \"rlib\"//" Cargo.toml && \
  # Skip executables
  sed -i'.bak' "/SCRIPTS = /d" Makefile.in && \
  ./configure --prefix=/usr/local --enable-static --disable-shared --disable-dependency-tracking \
    --disable-introspection --disable-tools --disable-pixbuf-loader --disable-nls --without-libiconv-prefix --without-libintl-prefix \
    ${DARWIN:+--disable-Bsymbolic} && \
  make install-strip && \
  rustup self uninstall -y && \
  cd ../ && rm -rf librsvg-${LIBRSVG_VERSION} && rm librsvg-${LIBRSVG_VERSION}.tar.gz
```

**Reply by kleisauke** [reply]
Should be easy, just swap `--enable-static --disable-shared` with `--enable-shared --disable-static`.
```diff
-  ./configure --prefix=/usr/local --enable-static --disable-shared --disable-dependency-tracking \
+  ./configure --prefix=/usr/local --enable-shared --disable-static --disable-dependency-tracking \
```

**Reply by adityapatadia** [reply]
Really noob here at libraries. Are there any advantages / disadvantages of using static vs shared libs? Your above recommendation worked fine.

**Reply by adityapatadia** [reply]
nvm, I have read online and enabled usage of shared libraries for all libs. Thanks a ton for your help :)

**Comment by adityapatadia** [comment]
@kleisauke I am facing another issue when trying to use sharp on nodejs. It says libvips.so.42 is not found. When I see build logs of libvips, it moves library libvips.so.42.15 to lib folder. Any idea how to resolve that?

**Reply by kleisauke** [reply]
Do you mean it installs the shared library in `/usr/local/lib/x86_64-linux-gnu` instead of `/usr/local/lib`? The former path is a debian-ism and can be avoided by configuring libvips with `--libdir=lib`, or you could set the `LD_LIBRARY_PATH=/usr/local/lib/x86_64-linux-gnu` environment variable.

**Reply by adityapatadia** [reply]
yes it installs in `/usr/local/lib/x86_64-linux-gnu`. Let me set the path.

**Reply by adityapatadia** [reply]
Passing `--libdir` solved the issue. Thanks again!

**Comment by adityapatadia** [comment]
@kleisauke all issues were resolved and build went perfectly fine. When we try to deploy on production, the container which we built keeps crashing without any error logs. What is best way to debug what is going wrong?
